src/core/TESTING.md

```diff
@@ -1087,3 +1095,275 @@ describe('Plugin', () => {
+    };
+  }
+}
+```
+ 
+We could write a test that asserts the service properly emit processed results until an invalid data is encountered:
+ 
+```typescript
+it('processDataStream throw an error when processing invalid data', () => {
+  getTestScheduler().run(({ expectObservable, hot }) => {
+    const service = new BarService();
+    const { processDataStream } = service.setup();
+
+    const data = hot('--a--b--(c|)', {
+      a: { id: 'a' },
+      b: { id: 'invalid' },
+      c: { id: 'c' },
+    });
+
+    expectObservable(processDataStream(data)).toBe(
+      '--a--#',
+      {
+        a: { id: 'a', processed: 'additional-data' },
+      },
+      new ProcessError(`invalid data: 'invalid'`)
+    );
+  });
+});
+```
+
+Notes:
+ - the `-` symbol represents one virtual time frame.
+ - the `#` symbol represents an error.
+ - when throwing a base `Error`, the assertion must be against a string: `'[Error: {message}]'`
+ 
+#### Testing promise based observables
+
+In some cases, the observable we want to test is based on a Promise (like `of(somePromise).pipe(...)`). This can occur
+when using promise-based services, such as core's `http`, for instance.
+
+```typescript
+export const callServerAPI = (
+  http: HttpStart,
+  term: string,
+  { abort$ }: { abort$: Observable<undefined> }
+): Observable<SomeResponse> => {
+  let controller: AbortController | undefined;
+  if (abort$) {
+    controller = new AbortController();
+    abort$.subscribe(() => {
+      controller!.abort();
+    });
+  }
+  return from(
+    http.post<SomeResponse>('/api/endpoint', {
+      body: JSON.stringify({ term }),
+      signal: controller?.signal,
+    })
+  ).pipe(
+    takeUntil(abort$ ?? EMPTY),
+    map((response) => response.results)
+  );
+};
+```
+
+Testing that kind of promise based observable does not work out of the box with marble testing, as the asynchronous promise resolution
+is not handled by the test scheduler's 'sandbox'.
+
+Fortunately, there are workarounds for this problem. The most common one being to mock the promise-returning API to return
+an observable instead for testing, as `of(observable)` also works and returns the input observable. 
+
+Note that when doing so, the test suite must also include tests using a real promise value to ensure correct behavior in real situation.
+
+```typescript
+
+// NOTE: test scheduler do not properly work with promises because of their asynchronous nature.
+// we are cheating here by having `http.post` return an observable instead of a promise.
+// this still allows more finely grained testing about timing, and asserting that the method
+// works properly when `post` returns a real promise is handled in other tests of this suite
+
+it('callServerAPI result observable emits when the response is received', () => {
+  const http = httpServiceMock.createStartContract();
+  getTestScheduler().run(({ expectObservable, hot }) => {
+    http.post.mockReturnValue(hot('---(a|)', { a: { someData: 'foo' } }) as any);
```

Comment:
as `SomeResponse`  from https://github.com/elastic/kibana/pull/68749/files#diff-2560d3ff7063db0bd43bde7f0b1dd530R1320 ?

Reply:
It's `hot('---(a|)', { a: { someData: 'foo' } })` that is casted to `any` as it doesn't match the expected input type for `http.post.mockReturnValue` 

Maybe I should move the comment to this specific line if it's unclear?

Reply:
That's confusing for sure 😕